core/src/main/java/org/apache/spark/shuffle/api/metadata/ShuffleOutputTracker.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle.api.metadata;
+
+/**
+ * :: Private ::
+ *
+ * A plugin that can monitor the storage of shuffle data from map tasks, and can provide
+ * metadata to shuffle readers to aid their reading of shuffle blocks in reduce tasks.
+ * <p>
+ * {@link MapOutputMetadata} instances provided from the plugin tree's implementation of
+ * {@link org.apache.spark.shuffle.api.ShuffleMapOutputWriter} are sent to
+ * <p>
+ * Implementations MUST be thread-safe. Spark will invoke methods in this module in parallel.
```

Comment:
Add a note about what locking semantics are when the methods are invoked ?

Reply:
I'm not sure what can be added. Can you give an example of guidance we could give to developers?

Reply:
As an example, `registerShuffle` is invoked with write lock held, while `unregisterShuffle` is invoked without any lock.
As a public spi we are exposing, the locking semantics needs to be elaborated; so that users can evaluate what can be done safely/without contention, and what needs additional protection (for example, `updateMapOutput` can be performed without contention from `registerShuffle`)

Additionally, it also informs users about how to design their implementations. For example, if an implementation makes rpc calls (or other blocking calls) from `registerMapOutput` , there could be nontrivial side effects in driver.

(I just picked some api calls to illustrate, we need to elaborate for each).

Reply:
Is it sufficient to simply say that all APIs need to have exclusive access from each other, at least locking on the shuffle id level?

Reply:
`unregisterShuffle` currently does not lock ?

Reply:
Again, that's up to the implementation to handle, no?

Reply:
I was not referring to how implementations manage their state/locking (if they have additional state/coordination required, you are right, it has to be handled there - state managed across shuffles as an example for some custom impl).

I was referring to what guarantees/expectations that implementations have from spark (`MapOutputTracker` currently) when these methods are invoked. We have to document this, so that implementation can be both aware of the MT-safety guarantees we provide (and at what granularity) - so that they can make reasonable assumptions about what can/cant be done in an implementation.

For example, as mentioned above, `registerShuffle` is invoked only with write lock is held for a particular shuffle. While `unregisterShuffle` does not have that guarantee.

Reply:
The whole point of just saying that implementations must be thread-safe is that implementations cannot make _any_ guarantees about how the caller will invoke this API. Even if we can say that the current implementation of `MapOutputTracker` would guarantee locking on per-shuffle-id, for example, that isn't an API guarantee we can realistically make since `MapOutputTracker` is a module internal to Spark.

I think saying that there is no guarantee about concurrent access to all methods suffices. But then that basically means "the implementation should be thread-safe". Is that reasonable, or do we need to be more granular than that?